junit5/src/main/java/org/jboss/weld/junit5/auto/WeldJunit5AutoExtension.java

```diff
@@ -48,17 +51,20 @@
 public class WeldJunit5AutoExtension extends WeldJunit5Extension {
 
     @Override
-    protected void weldInit(Object testInstance, ExtensionContext context, Weld weld, WeldInitiator.Builder weldInitiatorBuilder) {
+    protected void weldInit(ExtensionContext context, Weld weld, WeldInitiator.Builder weldInitiatorBuilder) {
 
-        Class<?> testClass = testInstance.getClass();
+        List<?> testInstances = context.getRequiredTestInstances().getAllInstances();
+        List<Class<?>> testClasses = testInstances.stream().map(Object::getClass).collect(Collectors.toList());
 
-        ClassScanning.scanForRequiredBeanClass(testClass, weld, getExplicitInjectionInfoFromStore(context));
+        ClassScanning.scanForRequiredBeanClass(testClasses, weld, getExplicitInjectionInfoFromStore(context));
 
-        weld.addBeanClass(testClass);
-        weld.addExtension(new TestInstanceInjectionExtension(testClass, testInstance));
+        weld.addBeanClasses(testClasses.stream().toArray(Class<?>[]::new));
+        weld.addExtension(new TestInstanceInjectionExtension(testInstances));
```

Comment:
What is the purpose of passing in the whole list here? Nested classes ultimately cannot be considered CDI beans and the extension won't do any good for them. Besides, nested classes will still be injected into via the inherited logic from `WeldJunit5Extension#startWeldContainerIfAppropriate`. Unless I am missing something?

Reply:
As a side note, for me it's not a convincing reason to ignore invalid beans again after Weld does that already.

After all I think the whole statement is wrong. I changed

https://github.com/philippkunz/weld-junit/blob/nested-readme/junit5/src/main/java/org/jboss/weld/junit5/auto/ClassScanning.java#L203

from

            if (hasBeanDefiningAnnotation(foundClass)) {

to

            if (hasBeanDefiningAnnotation(foundClass) || testClasses.contains(foundClass)) {

That's now certainly not what I initially intended as scope of the changes anymore and I really hope I don't stretch it too much but what good would it do really if test classes would have to be annotated with for example a scope while the lifecycle is controlled by JUnit? Wouldn't the tests that would fail with both

        weld.addBeanClasses(testClasses.stream().toArray(Class<?>[]::new));

present and

        || testClasses.contains(foundClass)

removed which are a few all of them been missing a bean defining annotation before in the first place? I was also not able to fix a few samples without also removing

    <T> void rewriteTestClassScope(@Observes ProcessAnnotatedType<T> pat, BeanManager beanManager) {

        if (testInstancesByClass.containsKey(pat.getAnnotatedType().getJavaClass())) {
            pat.configureAnnotatedType().add(SINGLETON_LITERAL);
        }

    }

from `TestInstanceInjectionExtension` or would get

> org.jboss.weld.exceptions.DefinitionException: WELD-000046: At most one scope may be specified on [EnhancedAnnotatedTypeImpl] public @ApplicationScoped @EnableAutoWeld @Singleton class org.jboss.weld.junit5.auto.nested.OuterTestClassAndBeanSameInstanceTest

or

> java.lang.IllegalStateException: WELD-ENV-000016: Missing beans.xml file in META-INF

Altogether I doubt that test classes should require a bean defining annotation but I admit I'm not a CDI expert and hope for your patience with me

Reply:
I guess I answered some of your questions in my other comment - https://github.com/weld/weld-junit/pull/104#discussion_r448296578

But back to my original question here - why do you pass all tests instances (including inner classes) into the extension? It has no purpose that I can see. We are only able to turn the outermost class into a bean and change its injection target which is what the extension does. If you pass in all classes, you achieve nothing, because the extension will perform more work and ultimately will discard those classes because they are not valid bean candidates. Therefore, IMO you only need the outermost class inside the extension.

Reply:
@philippkunz this question still stands. I don't see a reason to pass in collection of classes of only the outermost will ever be affected by it.

Reply:
Allow me to turn your question the other way round. Why should only the outer-most test class be passed to Weld after we know it will ignore inner ones anyway? I don't see a compelling reason for that either. I tend to perceive passing the outer-most test class only and not others as kind of a duplication. At the moment that makes no practical difference and may be more a matter of taste and. I really don't like

    weld.addBeanClasses(testClasses.get(0));

Reply:
Because this is it is generating an extra work for nothing? It's not Weld's choice to ignore it them, CDI spec requires that. It is also confusing in the code because it is not apparent to anyone unfamiliar with CDI internals that this happens.

Reply:
All right. I had a feeling that line was worth a second thought and now added it.